### ISSUE TITLE
Improve exportDOMDocument to work with full HTML documents

### ISSUE THREAD

**Issue by damienalexandre** [jolicode/JoliTypo#4]
To allow JoliTypo to fix complete HTML documents, with doctype & `body` tag, the `exportDOMDocument` must be rewritten.

The `loadDOMDocument` hack needs to be considered.

**Comment by lyrixx** [comment]
Revisiting this one in 2026, with concrete tests on the current `main` (PHP 8.4.24 / libxml 2.9.14 as on CI, and PHP 8.5.4 / libxml 2.15.2). The issue is still valid, and the actual behaviour is worse than the README note suggests.

**TL;DR: if we bump the minimum PHP version to 8.4, this issue and every hack around it get fixed at once with `\Dom\HTMLDocument`. That is by far the option I would push for.**

## What `fix()` does today on a full document

| Input | Output |
|---|---|
| `<!DOCTYPE html><html><head><title>T</title></head><body><p>…</p></body></html>` | `<p>…</p>` only. Doctype, `<html>`, the whole `<head>` (title, meta, link) and `<body>` are silently dropped. This is the documented case. |
| Same with `<html lang="fr">` or `<body class="home">` | **Corrupted HTML.** Doctype and opening tags are kept, the closing `</body></html>` are removed, and the internal `<?xml encoding="UTF-8">` hack plus the injected `<meta http-equiv>` leak into the output. |
| `<BODY>` in uppercase | The `<body>` prefix hack is case sensitive, so a second body is added and the `<title>` ends up inside the body. |
| A comment after `</body>` | `</body><!-- c --></html>` leaks into the output. |

The `lang` case is probably the most common real-world full document, and it produces broken markup.

Root cause: `exportDOMDocument()` is a regex over `saveHTML()` that only matches a bare `<html>` and a bare `<body>`, and `fixContentEncoding()` injects things (the xml PI, a meta, a `<body>` prefix) that rely on that regex to be stripped again.

## Option A: stay on `DOMDocument` (PHP >= 8.1)

A regex-free export is about 25 lines: detect a full document on the input (`/<(!doctype|html)[\s>]/i`), remove the `?xml` PI/comment node from the DOM, then `saveHTML()` the whole document, or serialize the children of `<body>` one by one for a fragment, with `LIBXML_HTML_NODEFDTD` so that no doctype is invented. I prototyped it: all the cases above come out right, and the only test that changes is `Html5Test::testFullPageMarkup`, which currently encodes the body-only behaviour.

But it runs into the real problem behind this issue: **the output encoding depends on the libxml version**, and it always has.

| libxml | Fragment | Full document |
|---|---|---|
| 2.9 (CI) | named entities (`&ldquo;`) | raw UTF-8, only because of the injected meta |
| 2.15 (local) | raw UTF-8 | raw UTF-8 |

That is why 20 tests already fail locally on a recent libxml. Serializing node by node produces raw UTF-8 on 2.9 as well, so 19 fixtures asserting entities would break on CI. Option A therefore also needs an explicit decision on the output encoding, fixture updates, and it keeps the three hacks in `fixContentEncoding()` alive.

## Option B: bump to PHP 8.4 and use `\Dom\HTMLDocument`. This fixes everything at once.

**This is the one I really want to insist on.** PHP 8.4 ships the new DOM extension with a real HTML5 parser (Lexbor) and its own HTML5 serializer. I tested it on the same cases on PHP 8.4.24 / libxml 2.9.14 and on PHP 8.5.4 / libxml 2.15.2: identical, correct output everywhere.

With `Dom\HTMLDocument::createFromString($html, LIBXML_NOERROR)` then `$doc->saveHtml()` for a document or `$doc->body->innerHTML` for a fragment, we get, in one move:

- full documents preserved as is (doctype, `<html lang>`, `<head>`, `<body class>`, trailing comments), which closes this issue;
- **no more `<?xml encoding="UTF-8">` PI hack**: the parser takes UTF-8 by design (the `mb_detect_encoding` conversion from #7 can stay, the PI goes);
- **no more `<body>` prefix hack** and its case-sensitivity bug;
- **no more injected `<meta http-equiv>`**;
- **no more regex on the serialized output**;
- **output that no longer depends on the libxml version**: raw UTF-8 everywhere, which also fixes the 20 tests currently failing on recent libxml;
- HTML5 elements (`<header>`, `<article>`, `<video>`, …) parsed natively instead of tolerated.

Every single hack in `loadDOMDocument()`, `fixContentEncoding()` and `exportDOMDocument()` exists to work around libxml's HTML4 parser. Switching parser removes the reason for all of them, instead of patching one more symptom. **One version bump, five hacks deleted, one 13-year-old issue closed, and a test suite that stops depending on the host's libxml.**

The cost is real but bounded:

- minimum PHP 8.4 (8.1 has been EOL since December 2025, 8.2 goes EOL in December 2026), so a 2.0 release;
- the `\DOMDocument` / `\DOMText` / `\DOMNode` type hints in `processDOM()`, `doFix()` and `StateNode` become `\Dom\Document` / `\Dom\Text` / `\Dom\Node`;
- `tagName` is uppercase for HTML elements in the new API, so the protected-tags check must use `localName`;
- fixtures updated from named entities to UTF-8 (they need it anyway, see above);
- `Html5Test::testFullPageMarkup` flipped to expect the full document, and the README note removed.

## Suggested plan

1. Decide between A and B. **My strong recommendation: B, as a 2.0.** Option A is one more patch on top of the hacks. Option B deletes the hacks.
2. Whatever the choice, make raw UTF-8 the documented output contract and note the change in the CHANGELOG.
3. Drop the `easy-pick` label: the encoding decision makes this less trivial than it looks.

**Comment by damienalexandre** [comment]
I agree with the plan B, bump PHP 8.4 and use `\Dom\HTMLDocument` 👍 

This justify a 2.0.

**Comment by lyrixx** [comment]
Actually, I'm not sure this needs a 2.0. I think we can do everything in 1.x

**Comment by damienalexandre** [comment]
It will depend on the output differences. Switching the dom parser will have an impact. Let's do the work and decide later.
